Add KadoComi (Comic Walker)

hatozuki-programmer opened the pull request:
Closes #5422 

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

hatozuki-programmer committed:
KadoComi (Comic Walker) extension

Initial working commit for KadoComi (Comic Walker)

hatozuki-programmer committed:
Cleanup

hatozuki-programmer committed:
Clean up duplicate code

hatozuki-programmer committed:
Remove unnecessary dependencies from build.gradle

hatozuki-programmer committed:
Use book cover for thumbnail if available

vetleledaal commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,367 @@
+            genre = genreName
+            status = when (serializationStatus.lowercase()) {
+                SERIALIZATION_ONGOING -> SManga.ONGOING
+                SERIALIZATION_UNKNOWN -> SManga.UNKNOWN
```
You can just inline the 2 strings.

vetleledaal commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,367 @@
+            list.add(tagName)
+        }
+
+        return list.joinToString(", ")
```
`, ` is the default

vetleledaal commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,367 @@
+        val fragment = chapter.url.substringAfterLast("#")
+        val params = fragment.split("&")
+        val workCode = params[0].split("=")[1]
+        val episodeCode = params[1].split("=")[1]
```
Use `chapter.url.toHttpUrlOrNull()` and get info from that, instead of manual string handling.

vetleledaal commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,367 @@
+            val drmHash = manuscript.getString("drmHash")
+            val imageUrl = "$drmImageUrl#$drmHash" // hacky workaround to get drmHash to interceptor
+
+            list.add(Page(i, "", imageUrl = imageUrl))
```
```suggestion
            list.add(Page(i, imageUrl = imageUrl))
```

vetleledaal commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,367 @@
+        for (i in result.indices) {
+            result[i] = imageByteArray[i] xor hashByteArray[i % hashByteArray.size]
+        }
+        return result
```
```suggestion
        return imageByteArray.mapIndexed { idx, byte ->
            byte xor hashByteArray[idx % hashByteArray.size]
        }.toByteArray()
```

hatozuki-programmer committed:
Additional code clean up

hatozuki-programmer committed:
Convert to using DTOs

hatozuki-programmer committed:
Remove unnecessary fragment handling

Fragment isn't sent to the server, so it's not necessary to do any handling here

hatozuki-programmer committed:
Use fragment from URL

hatozuki-programmer committed:
Fix author/artist names

hatozuki-programmer committed:
Fix thumbnails

Default of bookCover was interfering with thumbnail determination logic

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+    override fun mangaDetailsParse(response: Response): SManga {
+        val details = KadoComiWorkDto.let {
+            json.decodeFromString<KadoComiWorkDto>(response.body.string())
+        }
```
```suggestion
        val details = json.decodeFromString<KadoComiWorkDto>(response.body.string())
```

Same for the others too

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+        var mangaAuthor: String? = null
+        var mangaArtist: String? = null
+
+        details.work.authors!!.forEach {
```
```suggestion
        details.work.authors?.forEach {
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+            )
+        }
+
+        return list
```
```kt
        return details.latestEpisodes?.result.orEmpty().map { episode ->
            SChapter.create().apply {
               ....
            }
        }
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+            throw Exception("このチャプターは非公開です\nChapter is not available!")
+        }
+
+        return list
```
```kt
        val pages = viewer.manuscripts.mapIndexed { idx, manuscript ->
            Page(....))
        }

        if (pages.isEmpty()) {
            throw Exception(....)
        }

        return pages
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+    // ============================= Utilities ==============================
+
+    private fun getWorkCode(manga: SManga): String {
+        return manga.url.split("/").reversed().first()
```
```suggestion
        return manga.url.substringAfterLast("/")
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+            )
+        }
+
+        return list
```
```kt
    return results.result.map { 
        SManga.create().apply {
            ....
        }
    }
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComi.kt:
```diff
@@ -0,0 +1,319 @@
+    private fun getGenres(work: KadoComiWork): String {
+        val list: MutableList<String> = arrayListOf(work.genre!!.name, work.subGenre!!.name)
+        work.tags!!.forEach { list.add(it.name) }
+        return list.joinToString()
```
```suggestion
        return listOfNotNull(work.genre?.name, work.subGenre?.name)
        .plus(work.tags.orEmpty().map { it.name })
        .joinToString()
```

bapeey commented on src/ja/kadocomi/src/eu/kanade/tachiyomi/extension/ja/kadocomi/KadoComiDto.kt:
```diff
@@ -0,0 +1,71 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class KadoComiWorkDto(
```
```suggestion
class KadoComiWorkDto(
```

Remove the `data` keyword because it generates methods that aren’t being used

bapeey reviewed (commented):
Some minor adjustments

hatozuki-programmer committed:
Implement changes from feedback

hatozuki-programmer committed:
Remove unneeded data keyword
